File: app/src/test/java/eu/kanade/tachiyomi/ChapterRecognitionTest.java
package eu.kanade.tachiyomi;

import org.junit.Before;
import org.junit.Test;

import eu.kanade.tachiyomi.data.database.models.Chapter;
import eu.kanade.tachiyomi.data.database.models.Manga;
import eu.kanade.tachiyomi.util.ChapterRecognition;

import static org.assertj.core.api.Assertions.assertThat;


public class ChapterRecognitionTest {

    Manga randomManga;

    private Chapter createChapter(String title) {
        Chapter chapter = Chapter.create();
        chapter.name = title;
        return chapter;
    }

    @Before
    public void setUp() {
        randomManga = new Manga();
        randomManga.title = "Something";
    }

    @Test
    public void testWithOneDigit() {
        Chapter c = createChapter("Ch.3: Self-proclaimed Genius");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(3f);
    }

    @Test
    public void testWithVolumeBefore() {
        Chapter c = createChapter("Vol.1 Ch.4: Misrepresentation");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(4f);
    }

    @Test
    public void testWithVolumeAndVersionNumber() {
        Chapter c = createChapter("Vol.1 Ch.3 (v2) Read Online");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(3f);
    }

    @Test
    public void testWithVolumeAndNumberInTitle() {
        Chapter c = createChapter("Vol.15 Ch.90: Here Blooms the Daylily, Part 4");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(90f);
    }

    @Test
    public void testWithVolumeAndSpecialChapter() {
        Chapter c = createChapter("Vol.10 Ch.42.5: Homecoming (Beginning)");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(42.5f);
    }

    @Test
    public void testWithJustANumber() {
        Chapter c = createChapter("Homecoming (Beginning) 42");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(42f);
    }

    @Test
    public void testWithJustASpecialChapter() {
        Chapter c = createChapter("Homecoming (Beginning) 42.5");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(42.5f);
    }

    @Test
    public void testWithNumberinMangaTitle() {
        Chapter c = createChapter("3x3 Eyes 96");
        Manga m = new Manga();
        m.title = "3x3 Eyes";
        ChapterRecognition.parseChapterNumber(c, m);
        assertThat(c.chapter_number).isEqualTo(96f);
    }

    @Test
    public void testWithColonAtTheEnd() {
        Chapter c = createChapter("Chapter 5: 365 days");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(5f);
    }

    @Test
    public void testWithZeros() {
        Chapter c = createChapter("Vol.001 Ch.003: Kaguya Doesn't Know Much");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(3f);
    }

    @Test
    public void testRange() {
        Chapter c = createChapter("Ch.191-200 Read Online");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(191f);
    }

    @Test
    public void testWithKeywordChAtTheEndOfTheManga() {
        // It should be 567, not 67 (ch is a keyword to get the chapter number)
        Chapter c = createChapter("Bleach 567: Down With Snowwhite");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(567f);
    }

    @Test
    public void testWithVersionBefore() {
        // It should be 84, not 2084
        Chapter c = createChapter("Onepunch-Man Punch Ver002 084 : Creeping Darkness");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(84f);
    }

    @Test
    public void testWithVersionBeforeAndAnotherNumber() {
        Chapter c = createChapter("Onepunch-Man Punch Ver002 086 : Creeping Darkness [3]");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(86f);
    }

    @Test
    public void testWithVolumeAfterChapter() {
        Chapter c = createChapter("Solanin 028 Vol. 2");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(28f);
    }

    @Test
    public void testWithVolumeAttachedToChapter() {
        Chapter c = createChapter("Ansatsu Kyoushitsu 011v002: Assembly Time");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(11f);
    }

    @Test
    public void testWithNumberInChapterTitle() {
        Chapter c = createChapter("Ansatsu Kyoushitsu 099 Present Time - 2nd Hour");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(99f);
    }

    @Test
    public void testAlphaSubChapters() {
        Chapter c = createChapter("Asu No Yoichi 19a");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(19.1f);
        c = createChapter("Asu No Yoichi 19b");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(19.2f);
    }

    @Test
    public void testChapterWithArcNumber() {
        Chapter c = createChapter("Manga title 123 - Vol 016 Arc title 002");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(123f);
    }

    @Test
    public void testChapterWithChapterPrefixAfterPart() {
        Chapter c = createChapter("Tokyo ESP 027: Part 002: Chapter 001");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(027f);
    }

    @Test
    public void testUnparsable() {
        Chapter c = createChapter("Foo");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(-1f);
    }

    @Test
    public void testChapterWithTime() {
        Chapter c = createChapter("Fairy Tail 404: 00:00");
        ChapterRecognition.parseChapterNumber(c, randomManga);
        assertThat(c.chapter_number).isEqualTo(404f);
    }

    @Test
    public void testPlainNumberInTitle() {
        Chapter c = createChapter("Kuroko no Basket 002 Monday at 840 on the Rooftop");
        Manga manga = new Manga();
        manga.title = "Kuroko no Basket";
        ChapterRecognition.parseChapterNumber(c, manga);
        assertThat(c.chapter_number).isEqualTo(2f);
    }
}

File: domain/src/test/java/tachiyomi/domain/chapter/service/ChapterRecognitionTest.kt
package tachiyomi.domain.chapter.service

import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.parallel.Execution
import org.junit.jupiter.api.parallel.ExecutionMode

@Execution(ExecutionMode.CONCURRENT)
class ChapterRecognitionTest {

    @Test
    fun `Basic Ch prefix`() {
        val mangaTitle = "Mokushiroku Alice"

        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch.4: Misrepresentation", 4.0)
    }

    @Test
    fun `Basic Ch prefix with space after period`() {
        val mangaTitle = "Mokushiroku Alice"

        assertChapter(mangaTitle, "Mokushiroku Alice Vol. 1 Ch. 4: Misrepresentation", 4.0)
    }

    @Test
    fun `Basic Ch prefix with decimal`() {
        val mangaTitle = "Mokushiroku Alice"

        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch.4.1: Misrepresentation", 4.1)
        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch.4.4: Misrepresentation", 4.4)
    }

    @Test
    fun `Basic Ch prefix with alpha postfix`() {
        val mangaTitle = "Mokushiroku Alice"

        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch.4.a: Misrepresentation", 4.1)
        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch.4.b: Misrepresentation", 4.2)
        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch.4.extra: Misrepresentation", 4.99)
    }

    @Test
    fun `Name containing one number`() {
        val mangaTitle = "Bleach"

        assertChapter(mangaTitle, "Bleach 567 Down With Snowwhite", 567.0)
    }

    @Test
    fun `Name containing one number and decimal`() {
        val mangaTitle = "Bleach"

        assertChapter(mangaTitle, "Bleach 567.1 Down With Snowwhite", 567.1)
        assertChapter(mangaTitle, "Bleach 567.4 Down With Snowwhite", 567.4)
    }

    @Test
    fun `Name containing one number and alpha`() {
        val mangaTitle = "Bleach"

        assertChapter(mangaTitle, "Bleach 567.a Down With Snowwhite", 567.1)
        assertChapter(mangaTitle, "Bleach 567.b Down With Snowwhite", 567.2)
        assertChapter(mangaTitle, "Bleach 567.extra Down With Snowwhite", 567.99)
    }

    @Test
    fun `Chapter containing manga title and number`() {
        val mangaTitle = "Solanin"

        assertChapter(mangaTitle, "Solanin 028 Vol. 2", 28.0)
    }

    @Test
    fun `Chapter containing manga title and number decimal`() {
        val mangaTitle = "Solanin"

        assertChapter(mangaTitle, "Solanin 028.1 Vol. 2", 28.1)
        assertChapter(mangaTitle, "Solanin 028.4 Vol. 2", 28.4)
    }

    @Test
    fun `Chapter containing manga title and number alpha`() {
        val mangaTitle = "Solanin"

        assertChapter(mangaTitle, "Solanin 028.a Vol. 2", 28.1)
        assertChapter(mangaTitle, "Solanin 028.b Vol. 2", 28.2)
        assertChapter(mangaTitle, "Solanin 028.extra Vol. 2", 28.99)
    }

    @Test
    fun `Extreme case`() {
        val mangaTitle = "Onepunch-Man"

        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 028", 28.0)
    }

    @Test
    fun `Extreme case with decimal`() {
        val mangaTitle = "Onepunch-Man"

        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 028.1", 28.1)
        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 028.4", 28.4)
    }

    @Test
    fun `Extreme case with alpha`() {
        val mangaTitle = "Onepunch-Man"

        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 028.a", 28.1)
        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 028.b", 28.2)
        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 028.extra", 28.99)
    }

    @Test
    fun `Chapter containing dot v2`() {
        val mangaTitle = "random"

        assertChapter(mangaTitle, "Vol.1 Ch.5v.2: Alones", 5.0)
    }

    @Test
    fun `Number in manga title`() {
        val mangaTitle = "Ayame 14"

        assertChapter(mangaTitle, "Ayame 14 1 - The summer of 14", 1.0)
    }

    @Test
    fun `Space between ch x`() {
        val mangaTitle = "Mokushiroku Alice"

        assertChapter(mangaTitle, "Mokushiroku Alice Vol.1 Ch. 4: Misrepresentation", 4.0)
    }

    @Test
    fun `Chapter title with ch substring`() {
        val mangaTitle = "Ayame 14"

        assertChapter(mangaTitle, "Vol.1 Ch.1: March 25 (First Day Cohabiting)", 1.0)
    }

    @Test
    fun `Chapter containing multiple zeros`() {
        val mangaTitle = "random"

        assertChapter(mangaTitle, "Vol.001 Ch.003: Kaguya Doesn't Know Much", 3.0)
    }

    @Test
    fun `Chapter with version before number`() {
        val mangaTitle = "Onepunch-Man"

        assertChapter(mangaTitle, "Onepunch-Man Punch Ver002 086 : Creeping Darkness [3]", 86.0)
    }

    @Test
    fun `Version attached to chapter number`() {
        val mangaTitle = "Ansatsu Kyoushitsu"

        assertChapter(mangaTitle, "Ansatsu Kyoushitsu 011v002: Assembly Time", 11.0)
    }

    /**
     * Case where the chapter title contains the chapter
     * But wait it's not actual the chapter number.
     */
    @Test
    fun `Number after manga title with chapter in chapter title case`() {
        val mangaTitle = "Tokyo ESP"

        assertChapter(mangaTitle, "Tokyo ESP 027: Part 002: Chapter 001", 027.0)
    }

    /**
     * Case where the chapter title contains the unwanted tag
     * But follow by chapter number.
     */
    @Test
    fun `Number after unwanted tag`() {
        val mangaTitle = "One-punch Man"

        assertChapter(mangaTitle, "Mag Version 195.5", 195.5)
    }

    @Test
    fun `Unparseable chapter`() {
        val mangaTitle = "random"

        assertChapter(mangaTitle, "Foo", -1.0)
    }

    @Test
    fun `Chapter with time in title`() {
        val mangaTitle = "random"

        assertChapter(mangaTitle, "Fairy Tail 404: 00:00", 404.0)
    }

    @Test
    fun `Chapter with alpha without dot`() {
        val mangaTitle = "random"

        assertChapter(mangaTitle, "Asu No Yoichi 19a", 19.1)
    }

    @Test
    fun `Chapter title containing extra and vol`() {
        val mangaTitle = "Fairy Tail"

        assertChapter(mangaTitle, "Fairy Tail 404.extravol002", 404.99)
        assertChapter(mangaTitle, "Fairy Tail 404 extravol002", 404.99)
    }

    @Test
    fun `Chapter title containing omake (japanese extra) and vol`() {
        val mangaTitle = "Fairy Tail"

        assertChapter(mangaTitle, "Fairy Tail 404.omakevol002", 404.98)
        assertChapter(mangaTitle, "Fairy Tail 404 omakevol002", 404.98)
    }

    @Test
    fun `Chapter title containing special and vol`() {
        val mangaTitle = "Fairy Tail"

        assertChapter(mangaTitle, "Fairy Tail 404.specialvol002", 404.97)
        assertChapter(mangaTitle, "Fairy Tail 404 specialvol002", 404.97)
    }

    @Test
    fun `Chapter title containing commas`() {
        val mangaTitle = "One Piece"

        assertChapter(mangaTitle, "One Piece 300,a", 300.1)
        assertChapter(mangaTitle, "One Piece Ch,123,extra", 123.99)
        assertChapter(mangaTitle, "One Piece the sunny, goes swimming 024,005", 24.005)
    }

    @Test
    fun `Chapter title containing hyphens`() {
        val mangaTitle = "Solo Leveling"

        assertChapter(mangaTitle, "ch 122-a", 122.1)
        assertChapter(mangaTitle, "Solo Leveling Ch.123-extra", 123.99)
        assertChapter(mangaTitle, "Solo Leveling, 024-005", 24.005)
        assertChapter(mangaTitle, "Ch.191-200 Read Online", 191.200)
    }

    @Test
    fun `Chapters containing season`() {
        assertChapter("D.I.C.E", "D.I.C.E[Season 001] Ep. 007", 7.0)
    }

    @Test
    fun `Chapters in format sx - chapter xx`() {
        assertChapter("The Gamer", "S3 - Chapter 20", 20.0)
    }

    @Test
    fun `Chapters ending with s`() {
        assertChapter("One Outs", "One Outs 001", 1.0)
    }

    @Test
    fun `Chapters containing ordinals`() {
        val mangaTitle = "The Sister of the Woods with a Thousand Young"

        assertChapter(mangaTitle, "The 1st Night", 1.0)
        assertChapter(mangaTitle, "The 2nd Night", 2.0)
        assertChapter(mangaTitle, "The 3rd Night", 3.0)
        assertChapter(mangaTitle, "The 4th Night", 4.0)
    }

    private fun assertChapter(mangaTitle: String, name: String, expected: Double) {
        ChapterRecognition.parseChapterNumber(mangaTitle, name) shouldBe expected
    }
}
